### PR TITLE
Updates for GOES 16

### PR DESCRIPTION
**Pull request by jeanconn**
This 
 - [x] sets sets the previous GOES X-ray content to use the new GOES 16 X-ray data. 
 - [x] updates the GOES proton references to use the new MTA page and plots

For the GOES X-ray data, I didn't see a convenient way to just grab their plot, so I made a new one with the content.  The new content is fetched anyway to give timeline behavior somewhat consistent to the previous behavior.  

Output at https://cxc.cfa.harvard.edu/mta/ASPECT/arc_goes16_test/

This does not do anything about the HRC proxy.

### CODE REVIEW

**Comment by jeanconn** [comment]
Looks like this needs some more fixes

```
** ERROR - line 5: <<2020-Mar-25 00:11>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 939438 (char 939437)
** ERROR - line 5: <<2020-Mar-25 00:11>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 939438 (char 939437)
** ERROR - line 10: <<2020-Mar-25 02:36>> WARNING: /data/mta4/www/ace.html file contains 0 lines that start with AVERAGE (expect one)
** ERROR - line 15: <<2020-Mar-25 02:41>> WARNING: /data/mta4/www/ace.html file contains 0 lines that start with AVERAGE (expect one)
```

File: get_goes_x.py [hunk 1]
```diff
@@ -31,33 +33,32 @@
...
+shortdat = dat[dat['energy'] == '0.05-0.4nm']['flux', 'time_tag']
+shortdat.rename_column('flux', 'short')
+longdat = dat[dat['energy'] == '0.1-0.8nm']['flux', 'time_tag']
+longdat.rename_column('flux', 'long')
+# Join them on time (seems to be OK for these data)
+joindat = join(shortdat, longdat)
 
-secs = DateTime(mjd, format='mjd').secs
+# Add a time column with Chandra secs
+secs = []
```
Comment (taldcroft) [inline get_goes_x.py]: I think this can be a one-liner:
```
joindat['time'] = Time(joindat['time_tag']).cxcsec
```
Astropy time understands zulu time and CXC seconds.

File: plot_goes_x.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                    help='Plot file name')
+parser.add_argument('--h5',
+                    default='GOES_X_16.h5',
+                    help='HDF5 file name')
+args = parser.parse_args()
+
+
+
+
+h5 = tables.open_file(args.h5, mode='r')
```
Comment (taldcroft) [inline plot_goes_x.py]: Use the context manager `with tables.open_file.. as h5:`

File: plot_goes_x.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+table = h5.root.data[:]
+# Use just last 3 days if available
+table = table[table['time'] >= (DateTime() - 3).secs ]
+h5.close()
+
+
+plt.figure(1, figsize=(6, 4))
+for wave, color in zip(['long', 'short'], ['red', 'blue']):
+    vals = table[wave]
+    vals[vals <= 0] = 1e-10
```
Comment (taldcroft) [inline plot_goes_x.py]: Use numpy clip e.g..  `vals = vals.clip(min=1e-10)`.  If one cared about memory then the `out=` option is available (`vals.clip(min=1e-10, out=vals)`, but not needed here.

File: plot_goes_x.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    ticks, fig, ax = plot_cxctime(table['time'], vals, color=color, linewidth=.5,
+                                  label=f'GOES 16 {wave}')
+ax.set_ylim(1e-9, 1e-2)
```
Comment (taldcroft) [inline plot_goes_x.py]: You are 90% using the non-object matplotlib API, so just stick with `plt.ylim` and `plt.yscale`.  What you are currently doing by using the 2nd of two axes (`ax`) that are returned seems slightly ill-defined.  It works, but I think only accidentally because of how `plot_cxctime` is implemented.

File: plot_goes_x.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+# Use just last 3 days if available
+table = table[table['time'] >= (DateTime() - 3).secs ]
+h5.close()
+
+
+plt.figure(1, figsize=(6, 4))
+for wave, color in zip(['long', 'short'], ['red', 'blue']):
+    vals = table[wave]
+    vals[vals <= 0] = 1e-10
+    ticks, fig, ax = plot_cxctime(table['time'], vals, color=color, linewidth=.5,
```
Comment (taldcroft) [inline plot_goes_x.py]: Here you can ignore the return values (with the next comment).

File: plot_goes_x.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                                  label=f'GOES 16 {wave}')
```
Comment (taldcroft) [inline plot_goes_x.py]: Make the label be the actual wavelength range not short / long.

File: plot_goes_x.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+parser = argparse.ArgumentParser(description='')
+parser.add_argument('--out', type=str,
+                    default='goes_x.png',
+                    help='Plot file name')
+parser.add_argument('--h5',
+                    default='GOES_X_16.h5',
+                    help='HDF5 file name')
+args = parser.parse_args()
+
```
Comment (taldcroft) [inline plot_goes_x.py]: Only 1 vertical whitespace line here.  Your default code editor should be complaining at this.

File: plot_goes_x.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python
```
Comment (taldcroft) [inline plot_goes_x.py]: Include license and then a description of the program as a docstring.

File: plot_goes_x.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import argparse
```
Comment (taldcroft) [inline plot_goes_x.py]: System imports (argparse) in one block, 3rd party in the next block, then finally local package imports (not applicable here).

File: get_goes_x.py [hunk 10]
```diff
@@ -5,18 +5,20 @@
...
 
 import numpy as np
 from astropy.io import ascii
+from astropy.table import Table, join
 from Chandra.Time import DateTime
 
 parser = argparse.ArgumentParser(description='Get GOES_X data')
 parser.add_argument('--h5',
-                    default='GOES_X.h5',
+                    default='GOES_X_16.h5',
```
Comment (taldcroft) [inline get_goes_x.py]: Why this change to 16?  What happens when they go to 17?  Is the new file format incompatible with the existing GOES_X.h5?

File: get_goes_x.py [hunk 11]
```diff
@@ -31,33 +33,32 @@
...
           .format(err)))
     sys.exit(0)
 
-# Strip up to two rows at the end if any values are bad (i.e. negative)
-for _ in range(2):
-    if any(dat[name][-1] < 0 for name in data_colnames):
-        dat = dat[:-1]
-
-mjd = dat['mjd'] + dat['secs'] / 86400.
+# Select only the GOES 16 data (which is all there is)
```
Comment (taldcroft) [inline get_goes_x.py]: Satellite number should be a command line option so we can change in the future with only an update to task schedule.

File: get_goes_x.py [hunk 12]
```diff
@@ -31,33 +33,32 @@
...
+dat = dat[dat['satellite'] == 16]
+# Make a table for each of the two wavelengths
+shortdat = dat[dat['energy'] == '0.05-0.4nm']['flux', 'time_tag']
+shortdat.rename_column('flux', 'short')
+longdat = dat[dat['energy'] == '0.1-0.8nm']['flux', 'time_tag']
+longdat.rename_column('flux', 'long')
+# Join them on time (seems to be OK for these data)
```
Comment (taldcroft) [inline get_goes_x.py]: Can you add a warning if the two `time_tag` columns are not exactly equal?  With the `join` below you can silently be dropping data and never know.

File: get_goes_x.py [hunk 13]
```diff
@@ -31,33 +33,32 @@
...
-descrs = dat.dtype.descr
-descrs.append(('time', 'f8'))
-newdat = np.ndarray(len(dat), dtype=descrs)
-for colname in colnames:
-    newdat[colname] = dat[colname]
-newdat['time'] = secs
+# Save to h5
+newdat = joindat.as_array()
 
 h5 = tables.open_file(args.h5, mode='a',
```
Comment (taldcroft) [inline get_goes_x.py]: Change this to use a context manager to reflect current coding standards.

File: get_goes_x.py (hunk 10 above)
Comment (jeanconn) [reply get_goes_x.py]: I figured we were better off with a new file thanks to a totally different data source and different columns.

File: get_goes_x.py (hunk 11 above)
Comment (jeanconn) [reply get_goes_x.py]: That seemed like overconfigurating to me based on how not-robust the rest of this is.

File: arc3.cfg [hunk 14]
```diff
@@ -88,7 +88,7 @@ query_name = eclipse
   mta_ace_img  = https://cxc.harvard.edu/mta/mta_ace_plot.gif
   mta_goes_img = https://cxc.harvard.edu/mta/RADIATION/pgplot.gif
   mta_ace      = https://cxc.harvard.edu/mta/ace.html
-  mta_goes     = https://cxc.harvard.edu/mta/G13.html
+  mta_goes     = https://cxc.cfa.harvard.edu/mta/RADIATION_new/GOES/goes_pchan_p.html
```
[the remark this answers is not in the document]
Comment (jeanconn) [reply arc3.cfg]: Do you want me to ask MTA or did you want to do that?

File: plot_goes_x.py (hunk 7 above)
Comment (jeanconn) [reply plot_goes_x.py]: I'm in editor hell.

File: get_goes_x.py (hunk 12 above)
Comment (jeanconn) [reply get_goes_x.py]: That's a good idea, though without more information about the detectors I was inferring that the instrument just gives these two measurements all the time and the json file with separate records for them was just there to irritate me.  What are we going to do with an email warning about, say, missing data in one channel excluding data in the other channel?

File: get_goes_x.py (hunk 10 above)
Comment (jeanconn) [reply get_goes_x.py]: Do we have an expectation to use the GOES data saved in the replan central project for other analysis?  I couldn't figure out why it went back to like 2013 just to make the replan central plots, but if the point is that it should be a resource, that's a requirement I had not considered.

File: get_goes_x.py (hunk 1 above)
Comment (jeanconn) [reply get_goes_x.py]: Cool.  Thanks.

File: arc3.cfg (hunk 14 above)
[the remark this answers is not in the document]
Comment (jeanconn) [reply arc3.cfg]: Oh and they also have an HRC proxy column on that new page.  Do you want to use that in Replan Central now and let future discussion just change how the values are calculated for their page?

File: make_timeline.py [hunk 15]
```diff
@@ -198,7 +198,10 @@ def get_goes_x(tstart, tstop):
     p3 = h5.root.data.col('long')
     ok = (tstart < times) & (times < tstop)
     h5.close()
-    return times[ok], p3[ok]
+    if np.count_nonzero(ok) < 2:
+        return np.array([tstart, tstop]), np.array([1e-10, 1e-10])
+    else:
+        return times[ok], p3[ok]
```
Comment (jeanconn) [inline make_timeline.py]: I can also probably go ahead and change the name of this variable from p3 because I think it is not p3.

File: plot_goes_x.py (hunk 9 above)
Comment (jeanconn) [reply plot_goes_x.py]: Thanks!  And it is always hard to know how much time to take to improve code.  This looks like new code but is just the plot_hrc.py code because I didn't care and I didn't think it was worth generalizing.  Overall I probably should have just left this PR as  WIP while soak testing.

File: get_goes_x.py (hunk 12 above)
Comment (taldcroft) [reply get_goes_x.py]: We would be informed that data are being dropped from the file and would evaluate how much data are being dropped and whether a fix is needed, for instance interpolating across missing points.

File: get_goes_x.py (hunk 10 above)
Comment (taldcroft) [reply get_goes_x.py]: Yes it is a resource because AFAIK it is exceedingly difficult to get full-resolution archival GOES data.  So I decided to start collecting it locally.

File: get_goes_x.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeanconn) [reply get_goes_x.py]: I assume this feedback was actually for the *next* comment, as this one had a blank line above it (which you can't easily see with the delete edit above).

File: plot_goes_x.py (hunk 8 above)
Comment (jeanconn) [reply plot_goes_x.py]: Docstring or comment?  I assume we should just copy the standard LICENSE.rst into the project and comment all the files or some such?

File: arc3.cfg (hunk 14 above)
[the remark this answers is not in the document]
Comment (taldcroft) [reply arc3.cfg]: Anyway, it is what it is...

**Comment by taldcroft** [comment]
Let me know when this is ready for review again.

File: get_goes_x.py [hunk 16]
```diff
@@ -31,44 +41,63 @@
...
-newdat = np.ndarray(len(dat), dtype=descrs)
-for colname in colnames:
-    newdat[colname] = dat[colname]
-newdat['time'] = secs
+# Add a time column with Chandra secs and remove original time_tags
+times = Time(joindat['time_tag'])
+joindat['time'] = times.cxcsec
+joindat.remove_column('time_tag')
+
+# Add the other columns the old file format wanted
```
Comment (jeanconn) [inline get_goes_x.py]: I don't know if I should have just left these with placeholder values, but it seemed more work in the end to figure out what numbers to use if I didn't just put in values.

**Comment by jeanconn** [comment]
@taldcroft I think I've responded to your comments.  May have made a few things worse along the way!  Still not sure about the most trivial way to add another log axis in the X-plot to include the A, B, C, M, X labels we used to get.

File: get_goes_x.py (hunk 11 above)
Comment (jeanconn) [reply get_goes_x.py]: But I did it anyway.

**Comment by taldcroft** [comment]
About the A, B, C, M, X, those labels are actually fairly important.  I don't think you need a new axis or anything fancy, just use `text` to draw them at about the right place.

**Comment by jeanconn** [comment]
OK.  I'll get some labels.  Meanwhile, I'm still seeing some hdf5 errors in my on-the-side cron task

```
** ERROR - line 5: <<2020-Apr-01 13:13>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 940921 (char 940920)
** ERROR - line 5: <<2020-Apr-01 13:13>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 940921 (char 940920)
** ERROR - line 10: <<2020-Apr-01 21:48>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 941067 (char 941066)
** ERROR - line 10: <<2020-Apr-01 21:48>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 941067 (char 941066)
```

But it looks like it is still updating?  I'm not seeing a way to troubleshoot that.

File: get_goes_x.py [hunk 17]
```diff
@@ -31,44 +41,63 @@
...
     sys.exit(0)
 
-
-colnames = 'year month dom  hhmm  mjd secs short long ratio'.split()
-data_colnames = colnames[-3:]
-
 try:
-    dat = ascii.read(urldat, guess=False, Reader=ascii.NoHeader,
-                     data_start=3, names=colnames)
+    dat = Table(json.loads(urldat))
```
Comment (jeanconn) [inline get_goes_x.py:45]: Actually, my mistake, I thought the GOES_X warning was from the h5 code, but it looks like it is our own warning on a failure here.  So I'm guessing that 

```
** ERROR - line 5: <<2020-Apr-01 13:13>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 940921 (char 940920)
** ERROR - line 5: <<2020-Apr-01 13:13>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 940921 (char 940920)
** ERROR - line 10: <<2020-Apr-01 21:48>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 941067 (char 941066)
** ERROR - line 10: <<2020-Apr-01 21:48>> Warning: malformed GOES_X data so table read failed: Expecting ',' delimiter: line 1 column 941067 (char 941066)
```

is either a hiccup with the fetch or this is a race condition on the swpc webserver and we're occasionally fetching an incomplete or empty file.  I'm wondering about either printing out the data to check and also chucking the warning in the logs because I don't think we care.

Comment (taldcroft) [reply get_goes_x.py:45]: Maybe it would be worth putting this into the log, but without ERROR or warning.  This way we get some visibility into how often this is happening without spamming us with ignorable warnings.

**Comment by jeanconn** [comment]
Thanks.  And yes, I do.  Do you want it in the PR?

**Comment by taldcroft** [comment]
Maybe doesn't need to be in the PR, but perhaps as a gist and then put the gist link in the code and in the PR here as a comment.

**Comment by jeanconn** [comment]
Sorry, I had trouble with `text` working in some kind of reproducible way, so I did the "fancier" thing.  I'll see if I can simplify.
